README
DBIx-Log4perl version 0.10
==========================

DBIx-Log4perl is an extension for DBI to selectively log SQL, parameters,
result-sets, transactions etc to a Log::Log4perl handle.

You will need the DBI and Log::Log4perl modules first.

Please see the NOTES section in the pod for incompatibilities.

Please make sure you have an up to date Test::More before running the
tests. You will need at least 0.62.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

NOTE: To to a full test you need a working DBI and DBD with access to
a database where you can create a table. Once you have this set up
you should run perl Makefile.PL like this:

perl Makefile.PL --dsn=dsn --user=dbuser --password=dbpassword --table=table

where:

  dsn is a valid DBI connection dsn e.g. dbd::mysql::database
  dbuser is a valid user to access database
  dbpassword is the password for dbuser
    (may be omitted if not password required)
  table is the name of a table to create for testing
    (NOTE: this table will be dropped if it already exists)

DEPENDENCIES

This module requires these other modules and libraries:

  DBI 1.50
  Log::Log4perl 1.04
  Test::More 0.62

COPYRIGHT AND LICENCE

Copyright (C) 2006 by Martin J. Evans

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.7 or,
at your option, any later version of Perl 5 you may have available.




Changes
Revision history for Perl extension DBIx::Log4perl.

0.27  NNN NNNN NN NNNN

  [MISCELLANEOUS]

  Added repository to META file.

0.26  Wed July 18 2012

  [Enhancements]

  Add DBIX_L4P_LOG_SQL mask and changed DBIx::Log4perl so when
  DBIX_L4P_LOG_SQL | DBIX_L4P_LOG_DELAYBINDPARAM are used together you
  ONLY get SQL and parameters logged.

0.25  Thu July 5 2012

  [Bug Fixes]

  Fixed RT 78226 - clone method takes attribute hash but it was
  not passed on to DBI or logged.

0.24  Fri January 17 2012

  Added a pod entry for dbix_l4p_logmask which I'd accidentally
  omitted.

  Fixed bug in fetch when DBIX_L4P_LOG_OUTPUT was enabled which failed
  to pass the handle to the debug function - thanks to Jimmy Grimberg
  for noticing and giving me enough information to track it down.

0.23  Wed October 13 2010

  Minor tweak to simple test and lib to ensure temporary files are removed
  even if full test not run.

0.22  Mon July 5 2010

  Changed simple.t test to avoid using piped Log::Log4perl::Appender::File
  and use recreate instead.

  Fixed bug in fetchrow_hashref if DBIX_L4P_LOG_OUTPUT was set. Thanks
  to Jimmy Grimberg for spotting this.

0.21  Wed March 3 2010

  I did not have a good day yesterday - lesson to be learnt, do not
  release new versions to cpan late at night. Fixed bug introduced
  whilst attempting to fix rt 55143 (should only effect people using
  execute_array and tuple status).

  Minor pod formatting changes/fixes.

0.20  Tues March 2 2010

  Fix rt 55143 where execute_array was logging array tuple status at
  INFO level instead of DEBUG level. Thanks to Eric Zylberstejn for
  finding the problem and providing the fix.

  Add the DBIX_L4P_LOG_DELAYBINDPARAM constant to delay logging of
  bind_param until execute.

0.19  Tues February 9 2010

  0.18 introduced a bug which stopped the catching and reporting of
  raised errors in DBI.

0.18  Mon January 22 2010

  Minor speedups in bind_param, bind_param_inout and execute methods.
    Thanks to Devel::NYTProf.

  Minor speedups in _unseen_sth.
    Thanks to Devel::NYTProf.

  Fix rt 53755 (fetchrow_array ignoring calling context). Thanks to
  Bill Rios for spotting this and identifying the problem.

  Log attributes passed to the connect call.

  When outputting the result of a selectall_* it didn't show the
  connection number.

  Add support for clone.
  NOTE: This changes introduces incompatibilities from previous
        versions as all attribute names have changed to become lower
        case.

0.17  Wed December 9 2009

  Fixed problem with no logging output when upgrading to Log::Log4perl
  1.26 and Perl 5.10.0. Was setting the log level before calling
  get_logger.

0.16  Thur October 1 2009

  Fixed rt50134 reported with a fix by David Gasson. The get_info method
  was not returning the value it got from the driver.

0.15  Wed June 24 2009

  Added the DBIx_l4p_ignore_err_regexp attribute.

0.14  Tues March 24 2009

  Fixed a few small bugs in SYNOPSIS section.

  Added logging of CallBacks existence in the error handler
  log the db::STORE method (when we can)

  Small optimisation in the way _unseen_sth is used in st.pm.

  When outputting the ParamValues in the error handler they were
  sorted on parameter name which is confusing since the parameter 10
  (or :p10) comes after 1 (or :p1). Changed the output to name the
  parameter market as well.

0.13  Thur November 6 2008

  Fixed bug in last_insert_id which was missing the level argument
  leading to various errors when logging last_insert_id.

  After looking at DBIx::Log4perl usage from an application I use
  through Devel::NYTProf I have reduced the time spent in
  DBIx::Log4perl a little.

  Fixed a few typos in the pod - thanks to Will Sheppard.

  Some time in the past the connect log message lost the ',' between
  the DSN and the username.

  Many tweaks to the pod.

  Now sets Data::Dumper::Quotekeys to 0 to avoid quoting hash keys
  when using Data::Dumper for log output.

  The logging of execute method calls now includes the SQL if execute
  was called with parameters.

0.12  Thur July 17 2008

  Code added to 0.11 to show statement number does not handle the
  selectrow_xxx and selectall_xxx methods when called with an already
  prepared statement handle. It does now.

  DBIx::Log4perl::st::bind_param log output was missing a closing
  bracket and also missing the connection number.

  The DBIx::Log4perl error handler did not cope with the fact that
  some statements (those magicked into existence by DBD::Oracle using
  ORA_RSET i.e., a cursor returned from a function or procedure) do
  not have any recorded SQL. This led to use of unitialised value in
  DBIx::Log4perl::_error_handler.

  Fix finish method bug when used on a statement handle returned as
  an output bound parameter in DBD::Oracle. The log handle was not
  always set (see notes on similar fix in 0.11) which can lead to
  unitialised value errors.

0.11  Thur April 17 2008

  DBIx::Log4perl did not support the use of statement handles returned
  as output bound parameters in DBD::Oracle because it had not seen
  them created and hence had not injected the log handle into them.

  DBIx::Log4perl was changed to only retrieve Oracle dbms_output if
  the logging handle has debug enabled but unfortunately it was
  continuing to enable dbms_output regardless of logging mode. As a
  result, once dbms_output was enabled, the Oracle dbms_output buffer
  starts filling and once full your procedures die.

  DBIx::Log4perl no longer sets $Log::Log4perl::caller_depth to 2 in the
  BEGIN because the depth depends on the method you are in. Instead
  Log::Log4perl::caller_depth is set in each method.

  Database methods and statement methods are now logged with one or more

t/lib.pl
# $Id$
use Cwd;
use File::Spec;

my $logtmp1;
my $logtmp2;

sub get_config
{
    my @v;
    if (-f ($file = 't/dbixl4p.config')  ||
	-f ($file = '../dbixl4p.config') ||
	-f ($file = 'dbixl4p.config')) {
	open IN, "$file";
	while (<IN>) {
	    chomp;
	    if ($_ eq 'UNDEF') {
		push @v, undef;
	    } else {
		push @v, $_;
	    }
	}
    }
    return @v;
}
sub config
{
    my $td = File::Spec->tmpdir or die q/Can't find a temporary directory to use. Please set TMPDIR, TEMP or TMP environment variables to a valid cirectory/;

#####    open OUT, ">pipe.pl" or die "Failed to create pipe.pl - $!";
#####    print OUT "#!$^X\n";
#####    print OUT 'while (<STDIN>) {open OUT, ">>dbixl4p.log"; print OUT $_;close OUT;}';
#####    close OUT;
#####    chmod 0777, "pipe.pl";

    $logtmp1 = File::Spec->catfile($td, 'dbixroot.log');
    $logtmp2 = File::Spec->catfile($td, 'dbix.log');
    my $cwd = getcwd();
    my $pipe = File::Spec->catfile($cwd, "pipe.pl");
    my $loginit = qq(
log4perl.logger = FATAL, LOGFILE

# LOGFILE appender used by root (here)
# log anything at level ERROR and above to $logtmp1
log4perl.appender.LOGFILE=Log::Log4perl::Appender::File
log4perl.appender.LOGFILE.filename=$logtmp1
log4perl.appender.LOGFILE.mode=append
log4perl.appender.LOGFILE.utf8=1
log4perl.appender.LOGFILE.autoflush=1
log4perl.appender.LOGFILE.Threshold = ERROR
log4perl.appender.LOGFILE.layout=PatternLayout
log4perl.appender.LOGFILE.layout.ConversionPattern=[%r] %F %L %c - %m%n

# Define DBIx::Log4perl to output DEBUG and above msgs to
# $logtmp2 using the simple layout
log4perl.logger.DBIx.Log4perl=DEBUG, A1
log4perl.appender.A1=Log::Log4perl::Appender::File
log4perl.appender.A1.filename=$logtmp2
log4perl.appender.A1.mode=append
log4perl.appender.A1.recreate=1
log4perl.appender.A1.recreate_check_interval=0
log4perl.appender.A1.utf8=1
log4perl.appender.A1.autoflush=1
log4perl.appender.A1.layout=Log::Log4perl::Layout::SimpleLayout
		     );

    ok (Log::Log4perl->init(\$loginit), 'init config');

    ok (Log::Log4perl->get_logger('DBIx::Log4perl'), 'get log handle');

    return($logtmp1, $logtmp2);
}

sub check_log
{
    my ($s, $file) = @_;
    $$s = "";
    return 0 if (! -r $file);
    my ($dev,$ino,$mode,$nlink,$uid,$gid,$rdev,$size, $atime,$mtime,$ctime,$blksize,$blocks) = stat($file);
    return 0 if ($size <= 0);
    open IN, "<$file";
    while (<IN>) {$$s .= $_};
    close IN;
    unlink $file;
    diag($$s) if ($ENV{TEST_VERBOSE});
    return 1;
}
1;


lib/DBIx/Log4perl/db.pm
# $Id: db.pm 245 2006-07-25 14:20:59Z martin $
use strict;
use warnings;
use DBI;
use Log::Log4perl;
use Data::Dumper;

package DBIx::Log4perl::db;
@DBIx::Log4perl::db::ISA = qw(DBI::db DBIx::Log4perl);
use DBIx::Log4perl::Constants qw (:masks $LogMask);

# $_glogger is not relied upon - it is just a fallback
my $_glogger;

my $_counter;                   # to hold sub to count

BEGIN {
    my $x = sub {
        my $start = shift;
        return sub {$start++}};
    $_counter = &$x(0);         # used to count dbh connections
}


sub STORE{
    my $dbh = shift;
    my @args = @_;

    my $h = $dbh->{private_DBIx_Log4perl};
    # as we don't set private_DBIx_Log4perl until the connect method sometimes
    # $h will not be set
    $dbh->_dbix_l4p_debug($h, 2, "STORE($h->{dbh_no})", @args)
        if ($h && ($h->{logmask} & DBIX_L4P_LOG_INPUT));


    return $dbh->SUPER::STORE(@args);
}

sub get_info
{
    my ($dbh, @args) = @_;

    my $h = $dbh->{private_DBIx_Log4perl};
    my $value = $dbh->SUPER::get_info(@args);

    $dbh->_dbix_l4p_debug($h, 2, "get_info($h->{dbh_no})", @args, $value)
        if ($h->{logmask} & DBIX_L4P_LOG_INPUT);
    return $value;
}
sub prepare {
    my($dbh, @args) = @_;

    my $h = $dbh->{private_DBIx_Log4perl};
    my $ctr = $h->{new_stmt_no}(); # get a new unique stmt counter in this dbh
    if (($h->{logmask} & DBIX_L4P_LOG_INPUT) &&
            (caller !~ /^DBIx::Log4perl/o) &&
                (caller !~ /^DBD::/o)) { # e.g. from selectall_arrayref
        $dbh->_dbix_l4p_debug($h, 2, "prepare($h->{dbh_no}.$ctr)", $args[0]);
    }

    my $sth = $dbh->SUPER::prepare(@args);
    if ($sth) {
        $sth->{private_DBIx_Log4perl} = $h;
        $sth->{private_DBIx_st_no} = $ctr;
    }

    return $sth;
}

sub prepare_cached {
    my($dbh, @args) = @_;

    my $h = $dbh->{private_DBIx_Log4perl};
    my $ctr = $h->{new_stmt_no}();
    if (($h->{logmask} & DBIX_L4P_LOG_INPUT) &&
            (caller !~ /^DBIx::Log4perl/o) &&
                (caller !~ /^DBD::/o)) { # e.g. from selectall_arrayref
        $dbh->_dbix_l4p_debug($h, 2,
                              "prepare_cached($h->{dbh_no}.$ctr)", $args[0]);
    }

    my $sth = $dbh->SUPER::prepare_cached(@args);
    if ($sth) {
        $sth->{private_DBIx_Log4perl} = $h;
        $sth->{private_DBIx_st_no} = $ctr;
    }
    return $sth;
}

sub do {
    my ($dbh, @args) = @_;
    my $h = $dbh->{private_DBIx_Log4perl};

    $h->{Statement} = $args[0];
    $dbh->_dbix_l4p_debug($h, 2, "do($h->{dbh_no})", @args)
        if ($h->{logmask} & DBIX_L4P_LOG_INPUT);

    my $affected = $dbh->SUPER::do(@args);

    if (!defined($affected)) {
        $dbh->_dbix_l4p_error(2, 'do error for ', @args)
            if (($h->{logmask} & DBIX_L4P_LOG_ERRCAPTURE) &&
                    !($h->{logmask} & DBIX_L4P_LOG_INPUT)); # not already logged
    } elsif (defined($affected) && $affected eq '0E0' &&
                 ($h->{logmask} & DBIX_L4P_LOG_WARNINGS)) {
        $dbh->_dbix_l4p_warning(2, 'no effect from ', @args);
    } elsif (($affected ne '0E0') && ($h->{logmask} & DBIX_L4P_LOG_INPUT)) {
        $dbh->_dbix_l4p_debug($h, 2, "affected($h->{dbh_no})", $affected);
        $dbh->_dbix_l4p_debug($h, 2, "\t" . $dbh->SUPER::errstr)
            if (!defined($affected));
    }
    return $affected;
}

sub selectrow_array {
    my ($dbh, @args) = @_;

    my $h = $dbh->{private_DBIx_Log4perl};

    if ($h->{logmask} & DBIX_L4P_LOG_INPUT) {
        if ((scalar(@args) > 0) && (ref $args[0])) {
            $dbh->_dbix_l4p_debug($h,
                2,
                "selectrow_array($h->{dbh_no}." .
                    $args[0]->{private_DBIx_st_no} . ")", @args);
        } else {
            $dbh->_dbix_l4p_debug($h, 2,
                                  "selectrow_array($h->{dbh_no})", @args);
        }
    }

    if (wantarray) {
	my @ret = $dbh->SUPER::selectrow_array(@args);
	$dbh->_dbix_l4p_debug($h, 2, "result($h->{dbh_no})", @ret)
	  if ($h->{logmask} & DBIX_L4P_LOG_OUTPUT);
	return @ret;

    } else {
	my $ret = $dbh->SUPER::selectrow_array(@args);
	$dbh->_dbix_l4p_debug($h, 2, "result($h->{dbh_no})", $ret)
	  if ($h->{logmask} & DBIX_L4P_LOG_OUTPUT);
	return $ret;
    }
}

sub selectrow_arrayref {
    my ($dbh, @args) = @_;

    my $h = $dbh->{private_DBIx_Log4perl};

    if ($h->{logmask} & DBIX_L4P_LOG_INPUT) {
        if ((scalar(@args) > 0) && (ref $args[0])) {
            $dbh->_dbix_l4p_debug(
                $h, 2,
                "selectrow_arrayref($h->{dbh_no}." .
                    $args[0]->{private_DBIx_st_no} . ")", @args);
        } else {
            $dbh->_dbix_l4p_debug(
                $h, 2, "selectrow_arrayref($h->{dbh_no})", @args);
        }
    }

    my $ref = $dbh->SUPER::selectrow_arrayref(@args);
    $dbh->_dbix_l4p_debug($h, 2, "result($h->{dbh_no})", $ref)
      if ($h->{logmask} & DBIX_L4P_LOG_OUTPUT);
    return $ref;
}

sub selectrow_hashref {
    my ($dbh, @args) = @_;

    my $h = $dbh->{private_DBIx_Log4perl};

    if ($h->{logmask} & DBIX_L4P_LOG_INPUT) {
        if ((scalar(@args) > 0) && (ref $args[0])){
            $dbh->_dbix_l4p_debug(
                $h, 2,
                "selectrow_hashref($h->{dbh_no}." .
                    $args[0]->{private_DBIx_st_no} . ")", @args)
        } else {
            $dbh->_dbix_l4p_debug($h, 2,
                                  "selectrow_hashref($h->{dbh_no})", @args);
        }
    }

    my $ref = $dbh->SUPER::selectrow_hashref(@args);
    # no need to show result - fetch will do this
    return $ref;

}

sub selectall_arrayref {
    my ($dbh, @args) = @_;

    my $h = $dbh->{private_DBIx_Log4perl};
    if ($h->{logmask} & DBIX_L4P_LOG_INPUT) {
        if ((scalar(@args) > 0) && (ref $args[0])) {
            $dbh->_dbix_l4p_debug(
                $h, 2,
                "selectall_arrayref($h->{dbh_no}." .
                    $args[0]->{private_DBIx_st_no} . ")", @args);
        } else {
            $dbh->_dbix_l4p_debug(
                $h, 2, "selectall_arrayref($h->{dbh_no})", @args);
        }
    }

    my $ref = $dbh->SUPER::selectall_arrayref(@args);
    $dbh->_dbix_l4p_debug($h, 2, "result($h->{dbh_no})", $ref)
      if ($h->{logmask} & DBIX_L4P_LOG_OUTPUT);
    return $ref;
}

sub selectall_hashref {
    my ($dbh, @args) = @_;

    my $h = $dbh->{private_DBIx_Log4perl};
    if ($h->{logmask} & DBIX_L4P_LOG_INPUT) {
        if ((scalar(@args) > 0) && (ref $args[0])) {
            $dbh->_dbix_l4p_debug(
                $h, 2,
                "selectall_hashref($h->{dbh_no}." .
                    $args[0]->{private_DBIx_st_no} . ")", @args);
        } else {
            $dbh->_dbix_l4p_debug($h, 2,
                                  "selectall_hashref($h->{dbh_no})", @args);
        }
    }

    my $ref = $dbh->SUPER::selectall_hashref(@args);
    # no need to show result - fetch will do this
    return $ref;

}

sub _make_counter {
    my $start = shift;
    return sub {$start++}
};

sub connected {

    my ($dbh, $dsn, $user, $pass, $attr) = @_;

    my %h = ();
    $h{dbh_no} = &$_counter();
    $h{new_stmt_no} = _make_counter(0); # get a new stmt count for this dbh

    if ($attr) {
      # check we have not got dbix_l4p_init without dbix_l4p_log or vice versa
	my ($a, $b) = (exists($attr->{dbix_l4p_init}),
		       exists($attr->{dbix_l4p_class}));
	croak ('dbix_l4p_init specified without dbix_l4p_class or vice versa')
	  if (($a xor $b));
	# if passed a Log4perl log handle use that
	if (exists($attr->{dbix_l4p_logger})) {
	    $h{logger} = $attr->{dbix_l4p_logger};
	} elsif ($a && $b) {
	    Log::Log4perl->init($attr->{dbix_l4p_init});
	    $h{logger} = Log::Log4perl->get_logger($attr->{dbix_l4p_class});
	    $h{init} = $attr->{dbix_l4p_init};
	    $h{class} = $attr->{dbix_l4p_class};
	} else {
	    $h{logger} = Log::Log4perl->get_logger('DBIx::Log4perl');
	}
	# save log mask
	$h{logmask} = $attr->{dbix_l4p_logmask}
	  if (exists($attr->{dbix_l4p_logmask}));
        # save error regexp
        $h{err_regexp} = $attr->{dbix_l4p_ignore_err_regexp}
            if (exists($attr->{dbix_l4p_ignore_err_regexp}));
	# remove our attrs from connection attrs
	#delete $attr->{dbix_l4p_init};
	#delete $attr->{dbix_l4p_class};
	#delete $attr->{dbix_l4p_logger};
	#delete $attr->{dbix_l4p_logmask};
        #delete $attr->{dbix_l4p_ignore_err_regexp};
    }
    # take global log mask if non defined
    $h{logmask} = $LogMask unless (exists($h{logmask}));

    $h{logger} = Log::Log4perl->get_logger('DBIx::Log4perl')
        unless (exists($h{logger}));
    $_glogger = $h{logger};

    # make sure you don't change the depth before calling get_logger:
    local $Log::Log4perl::caller_depth = $Log::Log4perl::caller_depth + 4;

    $h{dbd_specific} = 0;
    $h{driver} = $dbh->{Driver}->{Name};

    $dbh->{private_DBIx_Log4perl} = \%h;

    if ($h{logmask} & DBIX_L4P_LOG_CONNECT) {
        local $Data::Dumper::Indent = 0;
	$h{logger}->debug(
            "connect($h{dbh_no}): " .
                (defined($dsn) ? $dsn : '') . ', ' .
                    (defined($user) ? $user : '') . ', ' .
            Data::Dumper->Dump([$attr], [qw(attr)]));
	no strict 'refs';
	my $v = "DBD::" . $dbh->{Driver}->{Name} . "::VERSION";
	$h{logger}->info("DBI: " . $DBI::VERSION,
			 ", DBIx::Log4perl: " . $DBIx::Log4perl::VERSION .
			   ", Driver: " . $h{driver} . "(" .
			     $$v . ")");
    }

    #
    # If capturing errors then save any error handler and set_err Handler
    # passed to us and replace with our own.
    #
    if ($h{logmask} & DBIX_L4P_LOG_ERRCAPTURE) {
	$h{HandleError} = $attr->{HandleError}
	    if (exists($attr->{HandleError}));
	$h{HandleSetErr} = $attr->{HandleSetErr}
	    if (exists($attr->{HandleSetErr}));
	$dbh->{HandleError} = \&_error_handler;
	$dbh->{HandleSetErr} = \&_set_err_handler;
    }
    return;

}
sub clone {
    my ($dbh, @args) = @_;

    my $h = $dbh->{private_DBIx_Log4perl};
    if ($h->{logmask} & DBIX_L4P_LOG_CONNECT) {
        $dbh->_dbix_l4p_debug($h, 2, "clone($h->{dbh_no})", @args);
    }

    return $dbh->SUPER::clone(@args);
}

sub disconnect {
    my $dbh = shift;

    if ($dbh) {
	my $h;
	eval {
	    # Avoid
	    # (in cleanup) Can't call method "FETCH" on an undefined value
	    $h = $dbh->{private_DBIx_Log4perl};
	};
	if (!$@ && $h && defined($h->{logger})) {
            if ($h->{logmask} & DBIX_L4P_LOG_CONNECT) {
                local $Log::Log4perl::caller_depth =
                    $Log::Log4perl::caller_depth + 2;
                $dbh->_dbix_l4p_debug($h, 2, "disconnect($h->{dbh_no})");
            }
	}
    }
    return $dbh->SUPER::disconnect;

}

sub begin_work {
    my $dbh = shift;
    my $h = $dbh->{private_DBIx_Log4perl};

    $dbh->_dbix_l4p_debug($h, 2, "start transaction($h->{dbh_no})")
        if ($h->{logmask} & DBIX_L4P_LOG_TXN);

    return $dbh->SUPER::begin_work;
}

sub rollback {
    my $dbh = shift;
    my $h = $dbh->{private_DBIx_Log4perl};

    $dbh->_dbix_l4p_debug($h, 2, "roll back($h->{dbh_no})")
        if ($h->{logmask} & DBIX_L4P_LOG_TXN);

    return $dbh->SUPER::rollback;
}

sub commit {
    my $dbh = shift;

    my $h = $dbh->{private_DBIx_Log4perl};
    $dbh->_dbix_l4p_debug($h, 2, "commit($h->{dbh_no})")
        if ($h->{logmask} & DBIX_L4P_LOG_TXN);

    return $dbh->SUPER::commit;
}

sub last_insert_id {
    my ($dbh, @args) = @_;
    my $h = $dbh->{private_DBIx_Log4perl};

    $dbh->_dbix_l4p_debug($h, 2,
	sub {Data::Dumper->Dump([\@args], ["last_insert_id($h->{dbh_no})"])})
      if ($h->{logmask} & DBIX_L4P_LOG_INPUT);

    my $ret = $dbh->SUPER::last_insert_id(@args);
    $dbh->_dbix_l4p_debug($h, 2, sub {"\t" . DBI::neat($ret)})
      if ($h->{logmask} & DBIX_L4P_LOG_INPUT);
    return $ret;
}


#
# Error handler to capture errors and log them
# Whatever, errors are passed on.
# if the user of DBIx::Log4perl passed in an error handler that is called
# before returning.
#
sub _error_handler {
    my ($msg, $handle, $method_ret) = @_;

    local $Log::Log4perl::caller_depth = $Log::Log4perl::caller_depth + 1;

    my $dbh = $handle;
    my $lh;
    my $h = $handle->{private_DBIx_Log4perl};
    my $out = '';

    $lh = $_glogger;
    $lh = $h->{logger} if ($h && exists($h->{logger}));
    return 0 if (!$lh);

    if ($h && exists($h->{err_regexp})) {
        if ($dbh->err =~ $h->{err_regexp}) {
            goto FINISH;
        }
    }
    # start with error message, state and err
    $out .=  '  ' . '=' x 60 . "\n  $msg\n";
    $out .= "err() = " . $handle->err . "\n";
    $out .= "state() = " . $handle->state . "\n";

    if ($DBI::lasth) {
	$out .= "  lasth type: $DBI::lasth->{Type}\n"
	    if ($DBI::lasth->{Type});
	$out .= "  lasth Statement ($DBI::lasth):\n    " .
	    "$DBI::lasth->{Statement}\n"
		if ($DBI::lasth->{Statement});
    }
    # get db handle if we have an st
    my $type = $handle->{Type};
    my $sql;
    if ($type eq 'st') {	# given statement handle
	$dbh = $handle->{Database};
	$sql = $handle->{Statement};
    } else {
	# given db handle
	# We've got other stmts under this db but we'll deal with those later
	$sql = 'Possible SQL: ';
	$sql .= "/$h->{Statement}/" if (exists($h->{Statement}));
	$sql .= "/$dbh->{Statement}/"
	  if ($dbh->{Statement} &&
		(exists($h->{Statement}) &&
		 ($dbh->{Statement} ne $h->{Statement})));
    }

    my $dbname = exists($dbh->{Name}) ? $dbh->{Name} : "";
    my $username = exists($dbh->{Username}) ? $dbh->{Username} : "";
    $out .= "  DB: $dbname, Username: $username\n";
    $out .= "  handle type: $type\n  SQL: " . DBI::neat($sql) . "\n";
    $out .= '  db Kids=' . $dbh->{Kids} .
	', ActiveKids=' . $dbh->{ActiveKids} . "\n";
    $out .= "  DB errstr: " . $handle->errstr . "\n"
	if ($handle->errstr && ($handle->errstr ne $msg));

    if (exists($h->{ParamValues}) && $h->{ParamValues}) {
	$out .= "  ParamValues captured in HandleSetErr:\n    ";
	foreach (sort keys %{$h->{ParamValues}}) {
	    $out .= "$_=" . DBI::neat($h->{ParamValues}->{$_}) . ",";
	}
	$out .= "\n";
    }
    if ($type eq 'st') {
	my $str = "";
	if ($handle->{ParamValues}) {
	    foreach (sort keys %{$handle->{ParamValues}}) {
		$str .= "$_=" . DBI::neat($handle->{ParamValues}->{$_}) . ",";
	    }
	}
	$out .= "  ParamValues: $str\n";
	$out .= "  " .
	  Data::Dumper->Dump([$handle->{ParamArrays}], ['ParamArrays'])
	      if ($handle->{ParamArrays});
    }
    my @substmts;
    # get list of statements under the db
    push @substmts, $_ for (grep { defined } @{$dbh->{ChildHandles}});
    $out .= "  " . scalar(@substmts) . " sub statements:\n";
    if (scalar(@substmts)) {
	foreach my $stmt (@substmts) {
	    $out .= "  stmt($stmt):\n";
	    $out .= '    SQL(' . $stmt->{Statement} . ")\n  "
		if ($stmt->{Statement} &&
		    (exists($h->{Statement}) &&
		     ($h->{Statement} ne $stmt->{Statement})));
	    if (exists($stmt->{ParamValues}) && $stmt->{ParamValues}) {
		$out .= '   Params(';
		foreach (sort keys %{$stmt->{ParamValues}}) {
		    $out .= "$_=" . DBI::neat($stmt->{ParamValues}->{$_}) . ",";
		}
		$out .= ")\n";
	    }
	}
    }

    if (exists($dbh->{Callbacks})) {
        $out .= "  Callbacks exist for " .
            join(",", keys(%{$dbh->{Callbacks}})) . "\n";
    }
    local $Carp::MaxArgLen = 256;
    $out .= "  " .Carp::longmess("DBI error trap");
    $out .= "  " . "=" x 60 . "\n";
    $lh->fatal($out);

  FINISH:
    if ($h && exists($h->{ErrorHandler})) {
      return $h->{ErrorHandler}($msg, $handle, $method_ret);
    } else {
      return 0;			# pass error on
    }
}

#
# set_err handler so we can capture ParamValues before a statement
# is destroyed.
# If the use of DBIx::Log4perl passed in an error handler that is
# called before returning.
#
sub _set_err_handler {
    my ($handle, $err, $errstr, $state, $method) = @_;

    # Capture ParamValues
    if ($handle) {
	my $h = $handle->{private_DBIx_Log4perl};
	$h->{ParamValues} = $handle->{ParamValues}
	    if (exists($handle->{ParamValues}));
	return $h->{HandleSetErr}(@_) if (exists($h->{HandleSetErr}));
    }
    return 0;
}


1;


lib/DBIx/Log4perl/Constants.pm
use strict;
use warnings;

package DBIx::Log4perl::Constants;
require Exporter;
our @ISA = qw(Exporter);

use constant DBIX_L4P_LOG_DEFAULT => 0xdd;
use constant DBIX_L4P_LOG_ALL => 0xffffff;
use constant DBIX_L4P_LOG_INPUT => 0x1;
use constant DBIX_L4P_LOG_OUTPUT => 0x2;
use constant DBIX_L4P_LOG_CONNECT => 0x4;
use constant DBIX_L4P_LOG_TXN => 0x8;
use constant DBIX_L4P_LOG_ERRCAPTURE => 0x10;
use constant DBIX_L4P_LOG_WARNINGS => 0x20;
use constant DBIX_L4P_LOG_ERRORS => 0x40;
use constant DBIX_L4P_LOG_DBDSPECIFIC => 0x80;
use constant DBIX_L4P_LOG_DELAYBINDPARAM => 0x100;

our $LogMask = DBIX_L4P_LOG_DEFAULT;

our @EXPORT = ();
our @EXPORT_OK = qw ($LogMask);
our @EXPORT_MASKS = qw(DBIX_L4P_LOG_DEFAULT
		       DBIX_L4P_LOG_ALL
		       DBIX_L4P_LOG_INPUT
		       DBIX_L4P_LOG_OUTPUT
		       DBIX_L4P_LOG_CONNECT
		       DBIX_L4P_LOG_TXN
		       DBIX_L4P_LOG_ERRCAPTURE
		       DBIX_L4P_LOG_WARNINGS
		       DBIX_L4P_LOG_ERRORS
		       DBIX_L4P_LOG_DBDSPECIFIC
                       DBIX_L4P_LOG_DELAYBINDPARAM
		     );
our %EXPORT_TAGS = (masks => \@EXPORT_MASKS);
Exporter::export_ok_tags('masks');


1;


Makefile.PL
use 5.008000;
use ExtUtils::MakeMaker;
use Getopt::Long();
use Data::Dumper;

use vars qw($opt);
$opt = {};

Getopt::Long::GetOptions(
    $opt,
    "help" => sub {Usage()},
    "dsn=s",
    "user=s",
    "password:s",
    "table=s",
    ) || die Usage();


my ($dsn, $user, $password, $table) = ('', '', 'UNDEF', '');

$dsn = $ENV{DBI_DSN} if (exists($ENV{DBI_DSN}));
$user = $ENV{DBI_USER} if (exists($ENV{DBI_USER}));
$password = $ENV{DBI_PASS} if (exists($ENV{DBI_PASS}));
$table = $ENV{DBI_TABLE} if (exists($ENV{DBI_TABLE}));

$dsn = $opt->{dsn} if (exists($opt->{dsn}));
$user = $opt->{user} if (exists($opt->{user}));
$password = $opt->{password} if (exists($opt->{password}));
$table = $opt->{table} if (exists($opt->{table}));

$table ||= 'drop_me_dbixl4p';

open OUT, ">dbixl4p.config";
print OUT "$dsn\n";
print OUT "$user\n";
print OUT "$password\n";
print OUT "$table\n";
close OUT;

warn "\nPlease NOTE\nYou need to set and export the DBI_* environment variables or supply suitable values on the command line (see Perl Makefile.PL help) to run a complete test\n\n";

# See lib/ExtUtils/MakeMaker.pm for details of how to influence
# the contents of the Makefile that is written.
WriteMakefile(
    NAME              => 'DBIx::Log4perl',
    VERSION_FROM      => 'lib/DBIx/Log4perl.pm', # finds $VERSION
    PREREQ_PM         => {Log::Log4perl => 1.04,
		          DBI => 1.50,
		          Test::More => 0.62},
    ($] >= 5.005 ?     ## Add these new keywords supported since 5.005
      (ABSTRACT_FROM  => 'lib/DBIx/Log4perl.pm', # retrieve abstract from module
       AUTHOR         => 'Martin J. Evans <mjevans@cpan.org>') : ()),
    NO_META => 1
);


sub Usage {
  print STDERR <<"USAGE";
Usage: perl $0 [options]

Possible options are:

  --dsn=<db>         The DBI DSN to pass to DBI->connect
  --user=<user>      Use the username <user> for running the test suite;
                     defaults to no username
  --password=<pwd>   Use the password <pwd> for running the test suite;
                     defaults to no password
  --table=<table>    A table to be used as a test table in the DB
                     defaults to dbixl4p
USAGE
  exit 1;
}
